src/pconn.cc

```diff
@@ -547,6 +578,18 @@ PconnPool::closeN(int n)
     }
 }
 
+void
+PconnPool::closeToPeer(const CachePeer * const peer)
+{
+    auto hid = table;
+    hash_first(hid);
+    debugs(48, 3, "open connections: " << count());
+    for (auto walker = hash_next(hid); walker; walker = hash_next(hid)) {
+        // may delete walker
```

Comment:
Possible walker removal should not disrupt the loop because `hash_next()` sets `hid->next` to point to the element next to `walker`. The removal is actually done in `IdleConnList` destructor by `hash_remove_link()`. In our case (`hid->next != hl`) this function searches for the given element in the corresponding bucket and 'removes' it by resetting the 'next' pointer (i.e., this looks trivial and there should be no additional overhead compared to a `close()` system call).

Reply:
```C++
for (auto walker = hash_next(hid); walker; walker = hash_next(hid)) {
    // from hash_remove_link():
    P = &(*P)->next;
    *P = hl->next; // modifies `next`
    // from hash_next_bucket() called from hash_remove_link():
    while (hid->next == nullptr && ++hid->current_slot < hid->size)
        hid->next = hid->buckets[hid->current_slot]; // modifies `next`
}
```

> hash_next() sets `hid->next` to point to the element next to walker.

Yes.

> hash_remove_link() [is] resetting the 'next' pointer

Yes, and the above combination concerns me. How can we be sure that our two nested loops work correctly together when both use the _same_ `hid->next` field? At this point, I am not claiming that there is a bug. I am only asking a question.

The existing PconnPool::closeN() code also has two nested loops, but it has a `current` reset hack that calls hash_first(), and cloneN() iteration condition is very different. This new PconnPool::closeToPeer() code lacks that hack and relies on hash_next() to stop iterating at the right moment.

I hope that there is a simple answer. The last thing we want is a lengthy "follow this pointer modified across these three `if` statements and two loops" kind of proof that no ordinary human can validate reliably!..

P.S. I agree that walker destruction is "safe" and that there is no rehashing going on here. In the worst case, hash_remove_link() does a linear search, but we are going to ignore that possibility for now.

Reply:
> The loop maintains an invariant: at each iteration ... hid->next points to the next element.

IMO, the outer loop itself does not maintain any clear/obvious invariants involving `hid->next` because the loop itself does not fully control `hid->next` -- the loop calls IdleConnList::closeAllTo() that is allowed to modify `hid->next` for as long as it is allowed to "call back" its `parent` methods.

I probably agree that current IdleConnList::closeAllTo() code happens to preserve `hid->next`, but that preservation is not guaranteed by anything. I can imagine a hash_table refactoring that would result in `hid->next` being reset when items are deleted.

I do not think we can fully address this concern in this PR, but I am suggesting to add a comment:

```suggestion
        // may delete current but preserves hash iterator (i.e. hid->next) that hash_next() has advanced already
```

Let's also rename `walker` to `current` to avoid an implication that we are deleting an iterator itself.

Reply:
Done.